File: timps-code/src/models/claude.ts
import type { Message, ModelProvider, StreamEvent, StreamOptions, ToolDefinition } from '../config/types.js';
import { parseSSE } from '../utils/utils.js';

export function createClaudeProvider(apiKey: string, modelId?: string): ModelProvider {
  const model = modelId || 'claude-sonnet-4-20250514';
  return {
    name: 'claude',
    model,
    supportsFunctionCalling: true,
    async *stream(messages, tools, options): AsyncGenerator<StreamEvent> {
      const sysMsg = messages.find(m => m.role === 'system');
      const body: Record<string, unknown> = {
        model,
        max_tokens: options?.maxTokens || 8192,
        stream: true,
        messages: convertMessages(messages),
      };
      if (sysMsg) body.system = sysMsg.content;
      if (options?.temperature !== undefined) body.temperature = options.temperature;
      if (tools.length > 0) {
        body.tools = tools.map(t => ({ name: t.name, description: t.description, input_schema: t.inputSchema }));
      }

      const res = await fetch('https://api.anthropic.com/v1/messages', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json', 'x-api-key': apiKey, 'anthropic-version': '2023-06-01' },
        body: JSON.stringify(body),
        signal: options?.signal,
      });

      if (!res.ok) {
        const err = await res.text();
        yield { type: 'error', message: `Claude API ${res.status}: ${err}` };
        return;
      }

      let currentToolId = '';
      let inputTokens = 0, outputTokens = 0;

      for await (const raw of parseSSE(res.body!)) {
        let data: Record<string, unknown> = {};
        try { data = JSON.parse(raw.data); } catch { continue; }
        const ev = raw.event || '';

        if (ev === 'content_block_start') {
          const block = data.content_block as any;
          if (block?.type === 'tool_use') { currentToolId = block.id; yield { type: 'tool_start', id: block.id, name: block.name }; }
          else if (block?.type === 'thinking') yield { type: 'thinking', content: '' };
        } else if (ev === 'content_block_delta') {
          const delta = data.delta as any;
          if (delta?.type === 'text_delta') yield { type: 'text', content: delta.text };
          else if (delta?.type === 'input_json_delta') yield { type: 'tool_delta', id: currentToolId, argumentsChunk: delta.partial_json };
          else if (delta?.type === 'thinking_delta') yield { type: 'thinking', content: delta.thinking };
        } else if (ev === 'content_block_stop') {
          if (currentToolId) { yield { type: 'tool_end', id: currentToolId }; currentToolId = ''; }
        } else if (ev === 'message_start') {
          const u = (data.message as any)?.usage; if (u) inputTokens = u.input_tokens || 0;
        } else if (ev === 'message_delta') {
          const u = (data as any).usage; if (u) outputTokens = u.output_tokens || 0;
        } else if (ev === 'message_stop') {
          yield { type: 'done', usage: { inputTokens, outputTokens } };
        } else if (ev === 'error') {
          yield { type: 'error', message: String((data.error as any)?.message || 'Stream error') };
        }
      }
    },
  };
}

function convertMessages(messages: Message[]): unknown[] {
  const result: unknown[] = [];
  for (const msg of messages) {
    if (msg.role === 'system') continue;
    if (msg.role === 'assistant' && msg.toolCalls?.length) {
      const content: unknown[] = [];
      if (msg.content) content.push({ type: 'text', text: msg.content });
      for (const tc of msg.toolCalls) content.push({ type: 'tool_use', id: tc.id, name: tc.name, input: tc.arguments });
      result.push({ role: 'assistant', content });
    } else if (msg.role === 'tool') {
      result.push({ role: 'user', content: [{ type: 'tool_result', tool_use_id: msg.toolCallId, content: msg.content }] });
    } else {
      result.push({ role: msg.role, content: msg.content });
    }
  }
  return result;
}

File: crates/timps-providers/src/claude.rs
//! Anthropic Claude provider (messages API with SSE).
use anyhow::Result;
use async_trait::async_trait;
use futures::StreamExt;
use reqwest::Client;
use serde_json::{json, Value};
use tokio::sync::mpsc;
use tokio_util::io::StreamReader;
use tokio::io::AsyncBufReadExt;
use crate::{Message, Provider, ProviderStream, Role, StreamEvent, ToolCall};

struct PendingTool {
    id: String,
    name: String,
    args_buf: String,
}

pub struct ClaudeProvider { api_key: String, model: String }

impl ClaudeProvider {
    pub fn from_env() -> Self {
        Self {
            api_key: std::env::var("ANTHROPIC_API_KEY").unwrap_or_default(),
            model: std::env::var("CLAUDE_MODEL")
                .unwrap_or_else(|_| "claude-sonnet-4-5".to_string()),
        }
    }

    /// Convert internal messages to Anthropic API format.
    ///
    /// Tool results must be `user` messages with `tool_result` content blocks.
    /// Since the `Message` type lacks `tool_use_id`, we assign sequential
    /// placeholder IDs so the API accepts the turn.
    fn to_claude_messages(messages: &[Message]) -> Vec<Value> {
        let mut idx = 0u32;
        messages.iter()
            .filter(|m| m.role != Role::System)
            .map(|m| match m.role {
                Role::User => json!({ "role": "user", "content": m.content }),
                Role::Assistant => json!({ "role": "assistant", "content": m.content }),
                Role::Tool => {
                    idx += 1;
                    json!({
                        "role": "user",
                        "content": [{
                            "type": "tool_result",
                            "tool_use_id": format!("toolu_placeholder_{idx}"),
                            "content": m.content,
                        }]
                    })
                }
                Role::System => unreachable!(), // filtered above
            })
            .collect()
    }
}

#[async_trait]
impl Provider for ClaudeProvider {
    fn name(&self) -> &str { "claude" }
    fn default_model(&self) -> &str { "claude-sonnet-4-5" }

    async fn complete(&self, system: &str, messages: &[Message], tools: &[Value]) -> Result<ProviderStream> {
        let (tx, rx) = mpsc::channel(64);
        let client = Client::new();
        let api_key = self.api_key.clone();
        let model = self.model.clone();
        let system = system.to_string();
        let msgs = Self::to_claude_messages(messages);
        let tools = tools.to_vec();

        tokio::spawn(async move {
            let mut body = json!({
                "model": model,
                "max_tokens": 8096,
                "system": system,
                "messages": msgs,
                "stream": true,
            });
            if !tools.is_empty() {
                body["tools"] = json!(tools);
            }

            let resp = match client
                .post("https://api.anthropic.com/v1/messages")
                .header("x-api-key", &api_key)
                .header("anthropic-version", "2023-06-01")
                .json(&body)
                .send().await {
                    Ok(r) => r,
                    Err(e) => { let _ = tx.send(StreamEvent::Error(e.to_string())).await; return; }
                };

            // Check HTTP status before reading body.
            let status = resp.status();
            if !status.is_success() {
                let text = resp.text().await.unwrap_or_default();
                let detail = json!({ "status": status.as_u16(), "body": text });
                let _ = tx.send(StreamEvent::Error(detail.to_string())).await;
                let _ = tx.send(StreamEvent::Done).await;
                return;
            }

            // Stream SSE via line-delimited chunks.
            let byte_stream = resp.bytes_stream().map(|r| r.map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e)));
            let reader = StreamReader::new(byte_stream);
            let mut lines = reader.lines();

            let mut pending: Option<PendingTool> = None;

            loop {
                let line = match lines.next_line().await {
                    Ok(Some(l)) => l,
                    Ok(None) => break,
                    Err(e) => { let _ = tx.send(StreamEvent::Error(e.to_string())).await; break; }
                };

                let line = line.trim().to_string();
                if line.is_empty() { continue; }
                let data = match line.strip_prefix("data: ") {
                    Some(d) => d.to_string(),
                    None => continue,
                };

                let obj: Value = match serde_json::from_str(&data) {
                    Ok(v) => v,
                    Err(_) => continue,
                };

                match obj["type"].as_str() {
                    Some("content_block_start") => {
                        if obj["content_block"]["type"] == "tool_use" {
                            pending = Some(PendingTool {
                                id: obj["content_block"]["id"].as_str().unwrap_or("").to_string(),
                                name: obj["content_block"]["name"].as_str().unwrap_or("").to_string(),
                                args_buf: String::new(),
                            });
                        }
                    }
                    Some("content_block_delta") => {
                        if let Some(t) = obj["delta"]["text"].as_str() {
                            let _ = tx.send(StreamEvent::Token(t.to_string())).await;
                        }
                        if let Some(json_str) = obj["delta"]["input_json_delta"].as_str() {
                            if let Some(ref mut p) = pending {
                                p.args_buf.push_str(json_str);
                            }
                        }
                    }
                    Some("content_block_stop") => {
                        if let Some(p) = pending.take() {
                            let args: Value = serde_json::from_str(&p.args_buf).unwrap_or(Value::Null);
                            let _ = tx.send(StreamEvent::ToolCall(ToolCall {
                                id: p.id,
                                name: p.name,
                                args,
                            })).await;
                        }
                    }
                    Some("message_stop") => {
                        let _ = tx.send(StreamEvent::Done).await;
                        return;
                    }
                    Some("error") => {
                        let msg = obj["error"]["message"].as_str().unwrap_or("unknown error");
                        let _ = tx.send(StreamEvent::Error(msg.to_string())).await;
                        let _ = tx.send(StreamEvent::Done).await;
                        return;
                    }
                    _ => {}
                }
            }
            let _ = tx.send(StreamEvent::Done).await;
        });
        Ok(rx)
    }

    async fn list_models(&self) -> Result<Vec<String>> {
        Ok(vec![
            "claude-opus-4-5".into(), "claude-sonnet-4-5".into(),
            "claude-haiku-3-5".into(), "claude-3-opus-20240229".into(),
        ])
    }
}
